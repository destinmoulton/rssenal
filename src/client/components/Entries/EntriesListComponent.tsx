import { OrderedMap } from "immutable";
import * as React from "react";

import Entry from "./Entry/Entry";
import SettingsModalContainer from "../../containers/Modals/SettingsModalContainer";
import SortMenu from "./SortMenu";

import { propertyComparator } from "../../lib/sort";

import {
    IEntry,
    IFilter,
    TEntries,
    TEntryID,
    TFolders,
    TFeeds,
    ISetting
} from "../../interfaces";

export interface IEntriesListMapState {
    entries: TEntries;
    folders: TFolders;
    feeds: TFeeds;
    filter: IFilter;
    settings: ISetting[];
}

export interface IEntriesListMapDispatch {
    markEntryRead: (entry: IEntry) => void;
}

type TAllProps = IEntriesListMapState & IEntriesListMapDispatch;

interface IEntriesListState {
    sortBy: string;
    activeEntryId: string;
    currentTitle: string;
    processedEntries: OrderedMap<TEntryID, IEntry> | Iterable<IEntry>;
}

class EntriesListComponent extends React.Component<
    TAllProps,
    IEntriesListState
> {
    state = {
        sortBy: "publish_date:asc",
        activeEntryId: "",
        currentTitle: "",
        processedEntries: OrderedMap<TEntryID, IEntry>()
    };

    constructor(props: TAllProps) {
        super(props);

        // Setup the global/document level keypress events
        document.onkeydown = this._handleKeyDown.bind(this);
    }

    componentDidMount() {
        this._filterAndSortEntries(this.state.sortBy, false);
    }

    componentDidUpdate(prevProps: TAllProps) {
        if (this.props.filter.id !== prevProps.filter.id) {
            this._filterAndSortEntries(this.state.sortBy, true);

            // Reset scroll to top
            document.querySelector(".rss-entrylist-container").scrollTo(0, 0);

            this.setState({
                activeEntryId: ""
            });
        } else {
            this._filterAndSortEntries(this.state.sortBy, false);
        }
    }

    _filterAndSortEntries(sortBy: string, hasFilterChanged: boolean) {
        const { entries, folders, feeds, filter } = this.props;

        let filteredEntries = entries.toOrderedMap();

        let title = "All";

        switch (filter.limit) {
            case "feed":
                filteredEntries = filteredEntries
                    .filter(entry => {
                        return entry.feed_id === filter.id;
                    })
                    .toOrderedMap();

                const activeFeed = feeds.get(filter.id);

                title = activeFeed.title;
                break;
            case "folder":
                if (filter.id !== "all") {
                    const feedIds = feeds
                        .filter(feed => {
                            return feed.folder_id === filter.id;
                        })
                        .map(feed => {
                            return feed._id;
                        });

                    filteredEntries = entries
                        .filter(entry => {
                            return feedIds.includes(entry.feed_id);
                        })
                        .toOrderedMap();

                    const activeFolder = folders.get(filter.id);

                    title = activeFolder.name;
                }
                break;
        }

        let processedEntries = this._sortEntries(filteredEntries, sortBy);

        if (hasFilterChanged) {
            processedEntries = this._filterHiddenEntries(processedEntries);
        }

        this.setState({
            currentTitle: title,
            processedEntries: OrderedMap(processedEntries)
        });
    }

    _sortEntries(entries: TEntries, sortBy: string) {
        const sortParams = sortBy.split(":");
        return entries.sort((a, b) =>
            propertyComparator(a, b, sortParams[1], sortParams[0])
        );
    }

    _filterHiddenEntries(processedEntries: any) {
        const { settings } = this.props;

        if (false === settings[1].value) {
            return processedEntries.filter((entry: IEntry) => {
                return false === entry.has_read;
            });
        }
        return processedEntries;
    }

    _handleChangeSort = (e: any, data: any) => {
        this.setState({
            sortBy: data.value
        });

        this._filterAndSortEntries(data.value, true);
    };

    _handleKeyDown = (e: any) => {
        switch (e.key) {
            case "ArrowDown":
            case "j":
                //e.preventDefault();
                this._activateSiblingEntry("next");
                break;
            case "ArrowUp":
            case "k":
                //e.preventDefault();
                this._activateSiblingEntry("previous");
                break;
        }
    };

    _handleToggleEntry = (entryId: TEntryID) => {
        let nextActiveEntryId = entryId;
        if (this.state.activeEntryId === entryId) {
            nextActiveEntryId = "";
        }

        if (nextActiveEntryId !== "") {
            this._markRead(nextActiveEntryId);
        }

        this.setState({
            activeEntryId: nextActiveEntryId
        });
    };

    _activateSiblingEntry(direction: string) {
        const { activeEntryId, processedEntries } = this.state;

        let sibling: IEntry = null;

        if (activeEntryId === "") {
            sibling = processedEntries.first();
        } else {
            let previousEntry: IEntry = null;
            let nextEntry: IEntry = null;
            let found = false;
            processedEntries.map(entry => {
                if (found && !nextEntry) {
                    nextEntry = entry;
                }

                if (entry._id === activeEntryId) {
                    found = true;
                }

                if (!found) {
                    previousEntry = entry;
                }
            });

            if (direction === "next") {
                if (nextEntry !== null) {
                    sibling = nextEntry;
                } else {
                    sibling = processedEntries.last();
                }
            } else if (direction === "previous") {
                if (previousEntry !== null) {
                    sibling = previousEntry;
                } else {
                    sibling = processedEntries.first();
                }
            }
        }

        this._markRead(sibling._id);

        this._scrollToEntry(sibling._id);

        this.setState({
            activeEntryId: sibling._id
        });
    }

    _scrollToEntry(entryId: TEntryID) {
        const entryEl = document.getElementById("rss-entry-item-" + entryId);
        setTimeout(() => {
            document
                .querySelector(".rss-entrylist-container")
                .scrollTo(0, entryEl.offsetTop);
        }, 10);
    }

    _markRead(entryId: TEntryID) {
        const { entries, markEntryRead } = this.props;
        const entryToMark = entries.get(entryId);

        if (!entryToMark.has_read) {
            markEntryRead(entryToMark);
        }
    }

    _generateEntries() {
        const { activeEntryId, processedEntries } = this.state;
        const { settings } = this.props;

        let shouldShowImages = false;
        settings.forEach(setting => {
            if (setting.key === "show_images") {
                shouldShowImages = setting.value;
            }
        });

        return processedEntries.toArray().map(entry => {
            const isActive = entry._id === activeEntryId;
            return (
                <Entry
                    key={entry._id}
                    entry={entry}
                    toggleEntry={this._handleToggleEntry}
                    isActive={isActive}
                    shouldShowImages={shouldShowImages}
                />
            );
        });
    }

    render() {
        const { currentTitle, sortBy } = this.state;

        const entryList = this._generateEntries();

        return (
            <div>
                <div className="rss-entrylist-menu">
                    <div className="rss-entrylist-menu-title-container">
                        {currentTitle}
                    </div>
                    <div className="rss-entrylist-menu-sortselect-container">
                        <SortMenu
                            onChange={this._handleChangeSort}
                            currentSortBy={sortBy}
                        />&nbsp;&nbsp;
                        <SettingsModalContainer />
                    </div>
                </div>
                <div className="rss-entrylist-container">{entryList}</div>
            </div>
        );
    }
}

export default EntriesListComponent;
