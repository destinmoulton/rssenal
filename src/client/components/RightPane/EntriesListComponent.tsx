import { OrderedMap } from "immutable";
import * as React from "react";

import Entry from "./Entry/Entry";

import { propertyComparator } from "../../lib/sort";

import {
    IEntry,
    IFilter,
    TEntries,
    TEntryID,
    TFeeds,
    ISetting
} from "../../interfaces";

export interface IEntriesListMapState {
    entries: TEntries;
    feeds: TFeeds;
    filter: IFilter;
    settings: ISetting[];
}

export interface IEntriesListMapDispatch {
    markEntryRead: (entry: IEntry) => void;
}

interface IProps {
    sortBy: string;
}

type TAllProps = IEntriesListMapState & IEntriesListMapDispatch & IProps;

interface IEntriesListState {
    activeEntryId: string;
    currentTitle: string;
    visibleEntries: OrderedMap<TEntryID, IEntry> | Iterable<IEntry>;
}

class EntriesListComponent extends React.Component<
    TAllProps,
    IEntriesListState
> {
    state = {
        activeEntryId: "",
        currentTitle: "",
        visibleEntries: OrderedMap<TEntryID, IEntry>()
    };

    constructor(props: TAllProps) {
        super(props);

        // Setup the global/document level keypress events
        document.onkeydown = this._handleKeyDown.bind(this);
    }

    componentDidMount() {
        this._prepareVisibleEntries(false);
    }

    componentDidUpdate(prevProps: TAllProps) {
        if (this.props.filter.id !== prevProps.filter.id) {
            this._prepareVisibleEntries(true);

            // Reset scroll to top
            document.querySelector(".rss-entrylist-container").scrollTo(0, 0);

            this.setState({
                activeEntryId: ""
            });
        } else if (this.props.sortBy !== prevProps.sortBy) {
            this._prepareVisibleEntries(false);
        }
    }

    _prepareVisibleEntries(hasFilterChanged: boolean) {
        const filteredEntries = this._filterEntries();

        let visibleEntries = this._sortEntries(filteredEntries);

        if (hasFilterChanged) {
            visibleEntries = this._filterHiddenEntries(visibleEntries);
        }

        this.setState({
            visibleEntries: OrderedMap(visibleEntries)
        });
    }

    _filterEntries() {
        const { entries, feeds, filter } = this.props;

        switch (filter.limit) {
            case "feed":
                return entries
                    .filter(entry => {
                        return entry.feed_id === filter.id;
                    })
                    .toOrderedMap();
            case "folder":
                if (filter.id !== "all") {
                    const feedIds = feeds
                        .filter(feed => {
                            return feed.folder_id === filter.id;
                        })
                        .map(feed => {
                            return feed._id;
                        });

                    return entries
                        .filter(entry => {
                            return feedIds.includes(entry.feed_id);
                        })
                        .toOrderedMap();
                }
                break;
        }
        return entries.toOrderedMap();
    }

    _sortEntries(entries: TEntries) {
        const sortParams = this.props.sortBy.split(":");
        return entries.sort((a, b) =>
            propertyComparator(a, b, sortParams[1], sortParams[0])
        );
    }

    _filterHiddenEntries(visibleEntries: any) {
        const { settings } = this.props;

        if (false === settings[1].value) {
            return visibleEntries.filter((entry: IEntry) => {
                return false === entry.has_read;
            });
        }
        return visibleEntries;
    }

    _handleKeyDown = (e: any) => {
        switch (e.key) {
            case "ArrowDown":
            case "j":
                //e.preventDefault();
                this._activateSiblingEntry("next");
                break;
            case "ArrowUp":
            case "k":
                //e.preventDefault();
                this._activateSiblingEntry("previous");
                break;
        }
    };

    _handleToggleEntry = (entryId: TEntryID) => {
        let nextActiveEntryId = entryId;
        if (this.state.activeEntryId === entryId) {
            nextActiveEntryId = "";
        }

        if (nextActiveEntryId !== "") {
            this._markRead(nextActiveEntryId);
        }

        this.setState({
            activeEntryId: nextActiveEntryId
        });
    };

    _activateSiblingEntry(direction: string) {
        const { activeEntryId, visibleEntries } = this.state;

        let sibling: IEntry = null;

        if (activeEntryId === "") {
            sibling = visibleEntries.first();
        } else {
            let previousEntry: IEntry = null;
            let nextEntry: IEntry = null;
            let found = false;
            visibleEntries.map(entry => {
                if (found && !nextEntry) {
                    nextEntry = entry;
                }

                if (entry._id === activeEntryId) {
                    found = true;
                }

                if (!found) {
                    previousEntry = entry;
                }
            });

            if (direction === "next") {
                if (nextEntry !== null) {
                    sibling = nextEntry;
                } else {
                    sibling = visibleEntries.last();
                }
            } else if (direction === "previous") {
                if (previousEntry !== null) {
                    sibling = previousEntry;
                } else {
                    sibling = visibleEntries.first();
                }
            }
        }

        this._markRead(sibling._id);

        this._scrollToEntry(sibling._id);

        this.setState({
            activeEntryId: sibling._id
        });
    }

    _scrollToEntry(entryId: TEntryID) {
        const entryEl = document.getElementById("rss-entry-item-" + entryId);
        setTimeout(() => {
            document
                .querySelector(".rss-entrylist-container")
                .scrollTo(0, entryEl.offsetTop);
        }, 10);
    }

    _markRead(entryId: TEntryID) {
        const { entries, markEntryRead } = this.props;
        const entryToMark = entries.get(entryId);

        if (!entryToMark.has_read) {
            markEntryRead(entryToMark);
        }
    }

    _generateEntries() {
        const { activeEntryId, visibleEntries } = this.state;
        const { settings } = this.props;

        let shouldShowImages = false;
        settings.forEach(setting => {
            if (setting.key === "show_images") {
                shouldShowImages = setting.value;
            }
        });

        return visibleEntries.toArray().map(entry => {
            const isActive = entry._id === activeEntryId;
            return (
                <Entry
                    key={entry._id}
                    entry={entry}
                    toggleEntry={this._handleToggleEntry}
                    isActive={isActive}
                    shouldShowImages={shouldShowImages}
                />
            );
        });
    }

    render() {
        const entryList = this._generateEntries();

        return (
            <div>
                <div className="rss-entrylist-container">{entryList}</div>
            </div>
        );
    }
}

export default EntriesListComponent;
