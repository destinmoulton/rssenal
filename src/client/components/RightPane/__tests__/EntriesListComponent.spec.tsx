import * as React from "react";
import { shallow } from "enzyme";
import EntriesListComponent from "../EntriesListComponent";

import ENTRIES from "../../../../../test/data/immutable/ammendedEntries";
import FEEDS from "../../../../../test/data/immutable/feeds";
import SETTINGS_DATA from "../../../../../test/data/immutable/settings";
const FILTER = {
    limit: "all"
};

describe("<EntriesListComponent />", () => {
    it("renders and matches the snapshot", () => {
        const wrapper = shallow(
            <EntriesListComponent
                entries={ENTRIES}
                feeds={FEEDS}
                filter={FILTER}
                filteredEntries={ENTRIES}
                settings={SETTINGS_DATA}
                markEntryRead={jest.fn()}
                sortBy={"publish_date:asc"}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
