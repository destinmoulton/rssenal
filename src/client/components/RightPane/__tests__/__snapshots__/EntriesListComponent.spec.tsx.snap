// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntriesListComponent /> renders and matches the snapshot 1`] = `
<div
  className="rss-entrylist-container"
>
  <Entry
    entry={
      Object {
        "__v": 0,
        "_id": "5b33c76cb2438d5708dc1987",
        "content": "<p>Content for Open Source Database HA Resources from Severalnines</p>",
        "content_snippet": "<p>Content for Open Source Database HA Resources from Severalnines</p>",
        "creator": "Forrest Lymburner",
        "feedTitle": "High Scalability",
        "feed_id": "5b33c76cb2438d5708dc197e",
        "guid": "426227:4867632:36078004",
        "has_read": false,
        "link": "http://feedproxy.google.com/~r/HighScalability/~3/FFJQudxO2bU/open-source-database-ha-resources-from-severalnines.html",
        "publish_date": "2018-06-08T19:46:22.000Z",
        "timeAgo": "3 months ago",
        "title": "Open Source Database HA Resources from Severalnines",
      }
    }
    isActive={false}
    key="5b33c76cb2438d5708dc1987"
    shouldShowImages={true}
    toggleEntry={[Function]}
  />
  <Entry
    entry={
      Object {
        "__v": 0,
        "_id": "5b33c76cb2438d5708dc1985",
        "content": "<p>Content for June 15th, 2018</p>",
        "content_snippet": "<p>Content for June 15th, 2018</p>",
        "creator": "Todd Hoff",
        "feedTitle": "High Scalability",
        "feed_id": "5b33c76cb2438d5708dc197e",
        "guid": "426227:4867632:36079455",
        "has_read": false,
        "link": "http://feedproxy.google.com/~r/HighScalability/~3/sacuGsRa2G0/stuff-the-internet-says-on-scalability-for-june-15th-2018.html",
        "publish_date": "2018-06-15T15:57:00.000Z",
        "timeAgo": "3 months ago",
        "title": "Stuff The Internet Says On Scalability For June 15th, 2018",
      }
    }
    isActive={false}
    key="5b33c76cb2438d5708dc1985"
    shouldShowImages={true}
    toggleEntry={[Function]}
  />
  <Entry
    entry={
      Object {
        "__v": 0,
        "_id": "5b33c76cb2438d5708dc1983",
        "content": "<p>Content for How Ably Efficiently Implemented Consistent Hashing</p>",
        "content_snippet": "<p>Content for How Ably Efficiently Implemented Consistent Hashing</p>",
        "creator": "Todd Hoff",
        "feedTitle": "High Scalability",
        "feed_id": "5b33c76cb2438d5708dc197e",
        "guid": "426227:4867632:36081325",
        "has_read": false,
        "link": "http://feedproxy.google.com/~r/HighScalability/~3/EU6Dy48k59o/how-ably-efficiently-implemented-consistent-hashing.html",
        "publish_date": "2018-06-18T16:06:17.000Z",
        "timeAgo": "3 months ago",
        "title": "How Ably Efficiently Implemented Consistent Hashing",
      }
    }
    isActive={false}
    key="5b33c76cb2438d5708dc1983"
    shouldShowImages={true}
    toggleEntry={[Function]}
  />
  <Entry
    entry={
      Object {
        "__v": 0,
        "_id": "5b33c76cb2438d5708dc1981",
        "content": "<p>Content for Sponsored Post: Datadog, InMemory.Net, Triplebyte, Etleap, Scalyr, MemSQL</p>",
        "content_snippet": "<p>Content for Sponsored Post: Datadog, InMemory.Net, Triplebyte, Etleap, Scalyr, MemSQL</p>",
        "creator": "Todd Hoff",
        "feedTitle": "High Scalability",
        "feed_id": "5b33c76cb2438d5708dc197e",
        "guid": "426227:4867632:36081782",
        "has_read": false,
        "link": "http://feedproxy.google.com/~r/HighScalability/~3/gKe2fkOoT0I/sponsored-post-datadog-inmemorynet-triplebyte-etleap-scalyr.html",
        "publish_date": "2018-06-19T16:10:13.000Z",
        "timeAgo": "3 months ago",
        "title": "Sponsored Post: Datadog, InMemory.Net, Triplebyte, Etleap, Scalyr, MemSQL",
      }
    }
    isActive={false}
    key="5b33c76cb2438d5708dc1981"
    shouldShowImages={true}
    toggleEntry={[Function]}
  />
  <Entry
    entry={
      Object {
        "__v": 0,
        "_id": "5b33c76cb2438d5708dc197f",
        "content": "<p>Content for Stuff The Internet Says On Scalability For June 22nd, 2018</p>",
        "content_snippet": "<p>Content for Stuff The Internet Says On Scalability For June 22nd, 2018</p>",
        "creator": "Todd Hoff",
        "feedTitle": "High Scalability",
        "feed_id": "5b33c76cb2438d5708dc197e",
        "guid": "426227:4867632:36082622",
        "has_read": false,
        "link": "http://feedproxy.google.com/~r/HighScalability/~3/uqbrISxcA-E/stuff-the-internet-says-on-scalability-for-june-22nd-2018.html",
        "publish_date": "2018-06-22T15:56:46.000Z",
        "timeAgo": "3 months ago",
        "title": "Stuff The Internet Says On Scalability For June 22nd, 2018",
      }
    }
    isActive={false}
    key="5b33c76cb2438d5708dc197f"
    shouldShowImages={true}
    toggleEntry={[Function]}
  />
</div>
`;
